kernel/include/kswap.h

```diff
@@ -154,8 +156,12 @@ static ALWAYS_INLINE unsigned int do_swap(unsigned int key,
 		void *newsh = new_thread->switch_handle;
 
 		if (IS_ENABLED(CONFIG_SMP)) {
-			/* Active threads MUST have a null here */
+			/* Active threads must have a null here.  And
+			 * it must be seen before the scheduler lock
+			 * is released!
+			 */
 			new_thread->switch_handle = NULL;
+			barrier_dmem_fence_full(); /* write barrier */
```

Comment:
Hilariously, after I was the one who said we should just merge the full barrier API first for simplicity and worry about read/write (ARM, traditional) and/or acquire/release/consume (C++ terms, sigh) complexity later...

...I hit a spot where there's a clear use for finer barrier control within mere days.  Still don't think it's wrong to have the conservative API though.